components/tikv_util/src/worker/pool.rs

```diff
@@ -0,0 +1,427 @@
+// Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
+
+use prometheus::IntGauge;
+use std::error::Error;
+use std::fmt::{self, Debug, Display, Formatter};
+use std::sync::{Arc, Mutex};
+
+use futures::channel::mpsc::{unbounded, UnboundedReceiver, UnboundedSender};
+use futures::compat::Future01CompatExt;
+use futures::future::{self, FutureExt};
+use futures::stream::StreamExt;
+
+use super::metrics::*;
+use crate::future::poll_future_notify;
+use crate::timer::GLOBAL_TIMER_HANDLE;
+use crate::yatp_pool::{DefaultTicker, YatpPoolBuilder};
+use futures::executor::block_on;
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::time::{Duration, Instant};
+use yatp::{Remote, ThreadPool};
+
+#[derive(Eq, PartialEq)]
+pub enum ScheduleError<T> {
+    Stopped(T),
+    Full(T),
+}
+
+impl<T> ScheduleError<T> {
+    pub fn into_inner(self) -> T {
+        match self {
+            ScheduleError::Stopped(t) | ScheduleError::Full(t) => t,
+        }
+    }
+}
+
+impl<T> Error for ScheduleError<T> {}
+
+impl<T> Display for ScheduleError<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let msg = match *self {
+            ScheduleError::Stopped(_) => "channel has been closed",
+            ScheduleError::Full(_) => "channel is full",
+        };
+        write!(f, "{}", msg)
+    }
+}
+
+impl<T> Debug for ScheduleError<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Display::fmt(self, f)
+    }
+}
+
+pub trait Runnable: Send {
+    type Task: Display + Send + 'static;
+
+    /// Runs a task.
+    fn run(&mut self, _: Self::Task) {
+        unimplemented!()
+    }
+    fn on_tick(&mut self) {}
+    fn shutdown(&mut self) {}
+}
+
+pub trait RunnableWithTimer: Runnable {
+    fn on_timeout(&mut self);
+    fn get_interval(&self) -> Duration;
```

Comment:
Previously we can add multi timeout events for a runner. And, the runner can determine whether it's necessary to re-register the timer or not. Could you add these 2 features back?

Reply:
No. I think it is not necessary.  Because we only need one based tick event, and then we can trigger different event by this tick.

Reply:
Just like what I did in `region.rs`. We can write code as following:
```
struct Runner {
   event1_tick: usize,
   event2_tick: usize,
   trigger_event1_tick_limit: usize,
   trigger_event2_tick_limit: usize,
}

impl RunnableWithTimer for Runner {
   fn on_timeout(&mut self) {
       self.event1_tick += 1;
       if self.event1_tick > self.trigger_event1_tick_limit {
           self.on_event1_timeout(); 
       }
       .....
   }
}


```